### ISSUE TITLE
Kernel dies / disconnects after some idle time

### ISSUE THREAD

**Issue by mmagnuski** [spyder-ide/spyder-notebook#322]
## Description

After some time (between 15 minutes and a few hours) the notebook kernel dies/is disconnected and all variables are lost. This happens without any visible warning or error, the only change thatcan be seen in the notebook is the "No Kernel" information in top right of the notebook. Cells no longer can be evaluated (trying to evaluate a cell does not lead to any output or cell numeration appearing/incrementing). 
![no kernel](https://user-images.githubusercontent.com/8452354/95210904-3fdc2380-07ec-11eb-9f33-ac66262e6029.PNG)

I don't remember this happening with the previous version of spyder-notebook. I could leave the notebook open, and return to it after a few hours and it worked. And if some kernel death / disconnection did happen it was an exception while now it happens more often.
It is a bit problematic in my teaching setup, especially when students run computations that take some time and would like to return to the notebook and check the results. It seems that the longer the notebook runs (especially idle) the higher the chances of this problem (but this is only my hunch).
At least one time when it happened I was able open the notebook again and all the variables were there - so it seems that the kernel was only disconnected, not dead.

**What steps will reproduce the problem?**

So far I can only specify the conditions vaguely:

1. Open notebook
2. leave it open
3. check periodically if the problem appears

I'll update the steps when I learn something new.

## Versions and main components

* Spyder Version:  4.1.5
* Python Version: 3.7.3
* Qt Version: 5.9.7
* PyQt Version: 5.9.2 (pyqt5-sip=12.8.0)
* Operating system: Windows

### jupyter:
```
# Name                    Version                   Build  Channel
jupyter                   1.0.0                    py37_7
jupyter_client            6.1.6                      py_0
jupyter_console           6.2.0                      py_0
jupyter_core              4.6.3                    py37_0
jupyterlab                2.2.6                      py_0
jupyterlab_pygments       0.1.1                      py_0
jupyterlab_server         1.2.0                      py_0
notebook                  6.1.1                    py37_0
spyder-notebook           0.3.0                      py_0    spyder-ide
```

## Dependencies

```
# Mandatory:
atomicwrites >=1.2.0           :  1.4.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.6.0 (OK)
diff_match_patch >=20181111    :  20200713 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.13.0 (OK)
jedi =0.17.1                   :  0.17.1 (OK)
nbconvert >=4.0                :  6.0.6 (OK)
numpydoc >=0.6.0               :  0.9.1 (OK)
paramiko >=2.4.0               :  2.7.2 (OK)
parso =0.7.0                   :  0.7.0 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.6.3 (OK)
pygments >=2.0                 :  2.7.1 (OK)
pylint >=1.0                   :  2.6.0 (OK)
pyls >=0.34.0;<1.0.0           :  0.35.1 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  0.7.2 (OK)
qtconsole >=4.6.0              :  4.7.7 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
rtree >=0.8.3                  :  0.9.4 (OK)
sphinx >=0.6.6                 :  3.2.1 (OK)
spyder_kernels >=1.9.4;<1.10.0 :  1.9.4 (OK)
watchdog                       :  None (OK)
zmq >=17                       :  19.0.2 (OK)

# Optional:
cython >=0.21                  :  0.29.12 (OK)
matplotlib >=2.0.0             :  3.1.0 (OK)
numpy >=1.7                    :  1.18.1 (OK)
pandas >=0.13.1                :  0.25.3 (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  None (OK)

# Spyder plugins:
spyder_notebook                :  0.3.0 (OK)
```

**Comment by jitseniesen** [comment]
Hmm, I have not experienced this issue. If it happens again, can you please go to the plugin's option (hamburger) menu and look in the server log to see whether anything is reported there?

**Comment by mmagnuski** [comment]
I had this problem a few times recently but forgot to check the server info. But I have an interesting piece of information nevertheless.
Today this happened, I've restarted the notebook and it seemed to work well. But after I connected a console to the jupyter kernel (hamburger -> open console) it turned out the the variables and "cell number" in the console is different from what is in the notebook. It seems that the console was able to get hold of the jupyter session that the nb frontend lost. So it does not seem that the kernel dies but the notebook seems to be just unable to find it sometimes. Next time I have this issue I'll remember to check the "notebook server info".
